[WIP] Move title to header toolbar

vdwijngaert opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
This PR gives a go at moving the management of the entity title to the header toolbar area, as described in #27093.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

vdwijngaert committed:
Adds a dummy title component to the header toolbar

vdwijngaert committed:
Add panel component with control for editing entity title

paaljoachim commented:
This is what I see:

https://user-images.githubusercontent.com/5323259/117707651-14c91080-b1cf-11eb-97f1-10013372743b.mp4

Should the top bar title be centered in the top bar area?
The drop down text field box moves as one writes inside of it.

paaljoachim commented:
Hey Kerry @gwwar , Joen @jasmussen and George @Mamaduka 
It would be great with some advice/feedback to the PR.
Thank you!

Mamaduka commented:
Hi, Paal @paaljoachim 

As far as I can tell, this is still a work in progress. But will be happy to review/test once it reaches the first stable point.

vdwijngaert commented:
I have a lot of work on my plate the next couple of days, so I'll like be unable to resume work on this until at least Friday 14/5. Contributions, suggestions and/or advice on what's already here is welcome.

I think we could reuse the `<DocumentActions>` component instead of just copy-pasting it, like I did, but it's in the Site Editor package, and I'm not familiar with the policy on that (maybe export or move it somewhere?).

Mamaduka commented:
@vdwijngaert

Based on some recent issues and my general knowledge, duplication is sometimes better than early abstraction.

paaljoachim commented:
I'll ping a few folks who have worked/work on Full Site Editing who can give some advice.
@carolinan @Addison-Stavlo @aristath @jeyip

Addison-Stavlo commented:
> Based on some recent issues and my general knowledge, duplication is sometimes better than early abstraction.

^^ This.  We have heard this a lot when moving existing edit-post functionality to edit-site.  So I wouldn't worry too much about copying  the edit-site component code or moving it to another package to share.  Its probably best to just implement this in edit-post for now and we can always consolidate later if the need arises.

paaljoachim commented:
It would be great to get some movement on this PR!
Thanks!

paaljoachim commented:
This is what I see using Chrome and a local test site using Twenty Nineteen.

https://user-images.githubusercontent.com/5323259/138610767-a19be3bd-c326-450b-a96a-d1414e83d5fd.mp4

https://user-images.githubusercontent.com/5323259/138610772-d01957e8-4549-4a02-828d-0ba841b32ec4.mp4

vdwijngaert committed:
Divide post editor header in three parts

vdwijngaert committed:
Cleaned up edit post title dropdown

vdwijngaert committed:
Undo removing template title styles

vdwijngaert commented:
Status-update (see the issue):

> Okay guys, I'm back from my little Gutenberg-burnout, it seems :)
> 
> I resumed work on #31288, and it looks like it is functionally ready for part 1 as described by @paaljoachim. I updated the code to reflect the changes made to DocumentActions from the Site Editor package.
> 
> I also had to go ahead and update the layout of the header as well, to resemble the Site Editor header a bit more, otherwise the title would have always been a bit off-center. I still need to go trough the styling to remove some things that have been rendered obsolete because of this.

Obviously, having to duplicate code from some Site Editor components is less than ideal. In the future, it might make sense to have both editors use the same components for this, but I'm not familiar enough with the way that should work on an architectural level. If this is a priority, I'd like some feedback on that.

Like I said, I still need to go over some of the styling rules. Right now I've used some rules from the `DocumentActions` component in the `Post Title` component. As well as this, the header is now divided into three parts, much so like in the Site Editor, so the CSS for this needs to be sorted out as well. It now seems to work fine, but I've got a feeling there's a bit of redundant or obsolete styling rules now. Feedback and/or help on that is appreciated as well.


**Edit:** @paaljoachim, I fixed the _loading..._ issue as well.

paaljoachim commented:
This is a lot better @vdwijngaert Koen!

https://user-images.githubusercontent.com/5323259/138652287-f39f4977-4f74-4110-968e-21f33fba14d7.mp4

Mamaduka commented:
Based on the latest discussion in #27093. I think we can close this PR.

Thanks again for contributing, @vdwijngaert!

Mamaduka commented:
I'm going to close this PR.

Thanks again for your contribution, @vdwijngaert!
